Assets/Game/Scripts/MapGeneration/Algorithms/ArticulationPoint.cs

```diff
@@ -0,0 +1,116 @@
+    [SerializeField] private bool useRandomEndRow;
+
+    protected override void Start()
+    {
+        base.Start();
+    }
+
+    public override void GenerateMap(int width, int height, Random pseudoRandom)
+    {
+        int pathWidth = 1 + _mapManager.pathRadius * 2;
+        _mapManager.map = new int[width,height];
+        _mapManager.FillMap(1);
+        height /= pathWidth;
+        width /= pathWidth;
+        if (useRandomEndRow)
+        {
+            endRow = -1;
+        }
+        Dictionary<Coordinate, Coordinate> dfsPath = new Dictionary<Coordinate, Coordinate>();
+        if (startRow < 0 || startRow >= height || 
+            endRow < -1 || endRow >= height || 
+            startCol < 0 || startCol >= width || 
+            endCol < 0 || endCol >= width || 
+            endCol < startCol ||
+            (endCol - startCol) % 2 != 0)
+        {
+            return;
+        }
+        Coordinate startCoordinate = new Coordinate(startCol, startRow);
+        dfsPath[startCoordinate] = new Coordinate(startCol - 1, startRow);
+        Dfs(startCoordinate, pseudoRandom, ref dfsPath, endRow, height);
+        if (endRow == -1)
+        {
+            int maxLength = -1;
+            for (int i = startRow % 2; i < height; i += 2)
+            {
+                int length = GetPathLength(new Coordinate(endCol, i), dfsPath);
+                if (length > maxLength)
+                {
+                    maxLength = length;
+                    endRow = i;
+                }
+            }
+        }
+        Backtrack(new Coordinate(endCol, endRow), dfsPath);
+        _mapManager.CreatePassage(ScaledCoordinateToOriginal(new Coordinate(-1, startRow)), ScaledCoordinateToOriginal(new Coordinate(startCol, startRow)));
+        _mapManager.CreatePassage(ScaledCoordinateToOriginal(new Coordinate(width, endRow)), ScaledCoordinateToOriginal(new Coordinate(endCol, endRow)));
+    }
+    
+    private int GetPathLength(Coordinate current, Dictionary<Coordinate, Coordinate> dfsPath)
+    {
+        if (!dfsPath.ContainsKey(current))
+        {
+            return 0;
+        }
+        return 1 + GetPathLength(dfsPath[current], dfsPath);
+    }
+
+    private void Backtrack(Coordinate current, Dictionary<Coordinate, Coordinate> dfsPath)
+    {
+        if (!dfsPath.ContainsKey(current))
+        {
+            return;
+        }
+        _mapManager.CreatePassage(ScaledCoordinateToOriginal(current), ScaledCoordinateToOriginal(dfsPath[current]));
+        Backtrack(dfsPath[current], dfsPath);
+    }
+
+    private Coordinate ScaledCoordinateToOriginal(Coordinate coordinate)
+    {
+        return new Coordinate((_mapManager.pathRadius * 2 + 1) * coordinate.tileX + _mapManager.pathRadius,
+            (_mapManager.pathRadius * 2 + 1) * coordinate.tileY + _mapManager.pathRadius);
+    }
+
+    private void Dfs(Coordinate current, Random pseudoRandom, ref Dictionary<Coordinate, Coordinate> dfsPath, int endRow, int height)
```

Comment:
nicely done, things you may consider:
1) Your algo is limited to single mono-path where it does not allow cross path
2) You may want to also memorized how the path is generated, and have AI to consume it later rather having the AI to search for correct path.
3) Add a random seed option where you can always re-generate the same path for debugging

Reply:
1. The algorithm is design to be mono-path since we decided cross path may confuse the player.
2. I not sure about do we need AI to search path.
3. The random seed option is already implemented in MapManager.